src/node_file.cc

```diff
@@ -532,16 +533,17 @@ inline FSReqBase* AsyncDestCall(Environment* env,
     uv_fs_t* uv_req = req_wrap->req();
     uv_req->result = err;
     uv_req->path = nullptr;
-    after(uv_req);
+    after(uv_req);  // after may delete req_wrap if there is an error
     req_wrap = nullptr;
   }
 
   if (req_wrap != nullptr) {
-    args.GetReturnValue().Set(req_wrap->persistent());
+    req_wrap->SetReturnValue(args);
```

Comment:
Question: it is possible for `after` being called in another thread in the middle of this and make `req_wrap` a dangling pointer here? In my experiments if I move the check outside to the bindings, it seems possible.. Maybe I could set `uv_req->data = nullptr` in `FSReqAfterScope` (or `~ReqWrap` even?) and check `uv_req->data` instead to be safe... cc @bnoordhuis

Reply:
The after callback should always run on the main thread; only the work callback runs in a different thread.

Reply:
@bnoordhuis Just to make sure I am understanding correctly here..can I assume the `uv_fs_cb` passed to `uv_fs_*` will not get called if `uv_fs_*` returns an error right away?

Reply:
That's right.

Reply:
@bnoordhuis Thanks, I think I understand what's going on with `uv_fs_copyfile` keeping the loop open when it errors out with `EINVAL` now..this seems to fix the problem:

```diff
diff --git a/deps/uv/src/unix/fs.c b/deps/uv/src/unix/fs.c
index e0969a4c2f..f0446110f2 100644
--- a/deps/uv/src/unix/fs.c
+++ b/deps/uv/src/unix/fs.c
@@ -1513,8 +1513,11 @@ int uv_fs_copyfile(uv_loop_t* loop,
                    uv_fs_cb cb) {
   INIT(COPYFILE);

-  if (flags & ~UV_FS_COPYFILE_EXCL)
+  if (flags & ~UV_FS_COPYFILE_EXCL) {
+    if (cb != NULL)
+      uv__req_unregister(loop, req);
     return -EINVAL;
+  }

   PATH2;
   req->flags = flags;
```

Reply:
Ah, the bug is obvious in hindsight... the flags check should be done _before_ `INIT(COPYFILE)`.  Do you want to open a PR or should I do it?

Reply:
@bnoordhuis Yes, checking the flags before INIT seems more reasonable...would love to try to open a PR to libuv since I have not done that before. I will do that once I get to somewhere I can open my laptop.